[docs] Material-UI v4 is out 🎉 (#15766)
mui-org/material-ui: pages/blog/material-ui-v4-is-out.md

---
description: Material-UI v4 is out 🎉
---

# Material-UI v4 is out 🎉

> React components for faster and easier web development. Build your own design system, or start with Material Design.

Material-UI v4 has finally arrived. We are so excited about this release, as it defines better foundations for the UI components. Thank you to everyone, especially to [the team](/discover-more/team/), and to everyone who's contributed code, issue triage, and support. **Thank you.**

✨✨✨ See the **[4.0.0 Release Note](https://github.com/mui-org/material-ui/releases/tag/v4.0.0)** on GitHub. ✨✨✨

![TypeScript switch](/static/blog/material-ui-v4-is-out/banner.png)

<p class="blog-description">https://material-ui.com/</p>

## Summary

- [High-level goals for v4](#high-level-goals-for-v4)
- [What's new?](#whats-new)
- [What's next?](#whats-next)
- [Premium themes store](#premium-themes-store)

## High-level goals for v4

> Material-UI v1 was [released](https://medium.com/material-ui/material-ui-v1-is-out-e73ce13463eb) one year ago. This is our most important major release since then. For those unfamiliar with the history of the project, we didn't release a v2, and v3 only introduced a slight browsers support change.

Reviewer: ```suggestion
Material-UI v1 was [released](https://medium.com/material-ui/material-ui-v1-is-out-e73ce13463eb) one year ago. Version 4 is our most important major release since then. For those unfamiliar with the history of the project, we didn't release a v2, and v3 only introduced a slight browsers support change.
```

### Customization

The Material Design team at Google has made the customization of their design system framework a core feature: [Material Theming](https://material.io/design/material-theming/overview.html). It's an important dimension of the library for us. Since the release of v1, we have been improving the component customization demos, release after release. We demonstrate a wide range of different designs:

![Demo1](/static/blog/material-ui-v4-is-out/demo1.png)

<p class="blog-description"><a href="https://material-ui.com/components/text-fields/#customized-inputs">live demo</a></p>

![Demo2](/static/blog/material-ui-v4-is-out/demo2.png)

<p class="blog-description"><a href="https://material-ui.com/components/tabs/#customized-tabs">live demo</a></p>

![Demo3](/static/blog/material-ui-v4-is-out/demo3.png)

<p class="blog-description"><a href="https://mui-treasury.com/components/button#customized-button">live demo</a></p>

After listening to v3 customization experiences of hundreds of developers, we realized that we could do better. We have found 4 classes of common problems.

> - **CSS specificity**. The developers' style specificity needs to be higher to win over the style Material-UI injects in the page. By default, Material-UI injects its style at the end of the `<head>` element. But styled components, and other popular styling solutions, inject the style just before it, losing specificity. In order to solve the problem, we have introduced a new prop: `injectFirst`.

Reviewer: ```suggestion
- **CSS specificity**. The developers' style specificity needs to be higher to win over the style Material-UI injects in the page. By default, Material-UI injects its style at the end of the `<head>` element. However, styled components and other popular styling solutions inject the style just before it, losing specificity. To solve the problem, we have introduced a new prop: `injectFirst`.
```

<p class="blog-description">The DOM output once injectFirst is used.</p>

- **classes boilerplate**. Early in the v1 effort, we [decided](https://github.com/oliviertassinari/a-journey-toward-better-style) to use a CSS-in-JS styling solution: [JSS](https://cssinjs.org/). The large majority of the CSS-in-JS solutions output non-deterministic class names, e.g. `.fHmkjM`. This design decision helps the isolation of the style of each component, however, it makes the overrides harder. We introduced a `classes` API in v1 to target all our elements as an attempt to mitigate this problem.
We have observed the use of this API for months and have seen many people struggling with it. It can be challenging to apply the class name on the right element and requires boilerplate as well.
As an attempt to further improve the situation, we have changed the class name generation to [output global class names](/styles/advanced/#with-material-ui-core), while keeping the `classes` API working as before 💅.

![styled-components](/static/blog/material-ui-v4-is-out/styled-components.png)

<p class="blog-description">side by side comparison of both valid classes vs global approaches.</p>

>   ⚠️ Using global class names provide more power but comes with a responsibility. We encourage patterns that increase your custom style isolation.

Reviewer: `with a responsibility`?

```css
.MenuItem {
  color: black;
}
.MenuItem.Mui-selected { /* We increase the specificity */
  color: blue;
}
```
- **Dynamic variations.** As one of the most requested features, we have added the support for style functions:

```jsx
const useStyles = makeStyles({
  // style rule
  foo: props => ({
    backgroundColor: props.backgroundColor,
  }),
  bar: {
    // CSS property
    color: props => props.color,
  },
});

function MyComponent() {
  // Simulated props for the purpose of the example
  const props = { backgroundColor: 'black', color: 'white' };
  // Pass the props as the first argument of useStyles()
  const classes = useStyles(props);

  return <div className={`${classes.foo} ${classes.bar}`} />
}
```

### Documentation

> Documentation was reported as the 3rd most critical pain point in the developer survey. We have fixed some of the reported issues and resolved to continuously improve going forward.

Reviewer: ```suggestion
Documentation was reported as the 3rd most critical pain point in the developer survey. We have fixed some of the reported issues and aim to continuously improve going forward.
```

> - **TypeScript**. TypeScript's growth is impressive, the traffic of their documentation website has grown by a factor of 6 in 3 years. Material-UI v1 was released with built-in TypeScript definitions, but we needed to do more. Sebastian has lead the effort to migrate all the demos from JavaScript to TypeScript. This has two important implications. First, we type check our demos, it drastically improves our TypeScript test coverage. We have fixed many issues during the migration. Second, if you are writing your application with TypeScript, you can directly copy & paste our demos without needing to convert them, or having to fix the obscure errors.

Reviewer: ```suggestion
- **TypeScript**. TypeScript's growth is impressive, the traffic of their documentation website has grown by a factor of 6 in 3 years. Material-UI v1 was released with built-in TypeScript definitions, but we needed to do more. Sebastian has led the effort to migrate all the demos from JavaScript to TypeScript. This has two important implications. First, we type check our demos, it drastically improves our TypeScript test coverage. We have fixed many issues during the migration. Second, if you are writing your application with TypeScript, you can directly copy & paste our demos without needing to convert them, or having to fix the obscure errors.
```

> - **TypeScript**. TypeScript's growth is impressive, the traffic of their documentation website has grown by a factor of 6 in 3 years. Material-UI v1 was released with built-in TypeScript definitions, but we needed to do more. Sebastian has lead the effort to migrate all the demos from JavaScript to TypeScript. This has two important implications. First, we type check our demos, it drastically improves our TypeScript test coverage. We have fixed many issues during the migration. Second, if you are writing your application with TypeScript, you can directly copy & paste our demos without needing to convert them, or having to fix the obscure errors.

Reviewer: ```suggestion
- **TypeScript**. TypeScript's growth is impressive, the traffic of their documentation website has grown by a factor of 6 in 3 years. Material-UI v1 was released with built-in TypeScript definitions, but we needed to do more. Sebastian has lead the effort to migrate all the demos from JavaScript to TypeScript. This has two important implications. First, we type check our demos, this drastically improved our TypeScript test coverage. We have fixed many issues during the migration. Second, if you are writing your application with TypeScript, you can directly copy & paste our demos without needing to convert them, or having to fix the obscure errors.
```

![typescript](/static/blog/material-ui-v4-is-out/typescript.png)

<p class="blog-description">https://www.typescriptlang.org traffic estimation over time.</p>

![switch](/static/blog/material-ui-v4-is-out/switch.png)

<p class="blog-description">Use the JS/TS toggle to see code in JavaScript or TypeScript</p>

> - **i18n**. People come to the Material-UI's documentation from all around the world. We want to include as many people as possible 🌎🌍🌏. We have completed the effort started in v3 by working on the Algolia search support, Google search indexing, Table Of Contents and Side Nav infrastructure.

Reviewer: ```suggestion
- **i18n**. People come to Material-UI's documentation from all around the world. We want to include as many people as possible 🌎🌍🌏. We have completed the effort started in v3 by working on the Algolia search support, Google search indexing, Table Of Contents and Side Nav infrastructure.
```

  We would like to thank [Danica Shen](https://github.com/Domino987) and [Dominik Engel](https://github.com/Domino987) for their heroic work on the 🇨🇳 and 🇩🇪 translations, while not forgetting the other 348 (and growing) translators.
- **Best practices**. We are now recommending the use of the hooks API over the classes API wherever possible. We have migrated a large portion of the demos to showcase a single approach.
- **A better UX**. We have changed the menu organization to group all the components under a single navigation item. We have changed the background color to white to increase the text contrast and readability.

### Performance

You may be afraid that using Material-UI's components will bloat and slow down your website/application. Would you be better off writing your own components? Well, it's our mission to make this cost as minimal as possible 🚀.

> - **Tree shaking**. Material-UI v4 is the first version to support native tree shaking with ES modules. This has one important DX benefit—you can now us destructured imports when importing multiple components:

Reviewer 2: ```suggestion
- **Tree shaking**. Material-UI v4 is the first version to support native tree shaking with ES modules. This has one important DX benefit—you can now use destructured imports when importing multiple components:
```

```jsx
import { Button } from '@material-ui/core';

function MyButton() {
  const myRef = React.useRef();

  return <Button ref={myRef}>;
}
```
- **Hooks migration**. While there is [no plan](https://reactjs.org/docs/hooks-intro.html#gradual-adoption-strategy) to remove classes from React, the React Team [encourages](https://reactjs.org/docs/hooks-faq.html#do-i-need-to-rewrite-all-my-class-components) new code to be written with the hooks API.

>   Josh has lead an effort to rewrite the large majority of our components with the hooks API. The change has a couple of advantages.

Reviewer: ```suggestion
  Josh has led an effort to rewrite the vast majority of our components with the hooks API. The change has a couple of advantages.
```

  ![devtools](/static/blog/material-ui-v4-is-out/devtools.png)

  <p class="blog-description">React Dev Tools output for one Typography element in production (<a href="https://github.com/mui-org/material-ui/pull/15023">this POC</a> might further improve the situation).</p>

- **Concurrent mode**. React has shared its [release timeline](https://reactjs.org/blog/2018/11/27/react-16-roadmap.html#react-16x-q2-2019-the-one-with-concurrent-mode) for the important features coming into React. [Concurrent mode](https://reactjs.org/blog/2018/11/27/react-16-roadmap.html#react-16x-q2-2019-the-one-with-concurrent-mode) should be released within the next few months. Concurrent Mode lets React apps be more responsive by rendering component trees without blocking the main thread. We have worked on getting ready for it. Most of the work was related to fixing the [<React.StrictMode>](https://reactjs.org/docs/strict-mode.html) warnings. You should soon be able to leverage it 🔥.
- **Shallow tests.** We have removed the majority of our internal shallow tests to rely on [full mount tests](https://twitter.com/dan_abramov/status/1109461529227866112).

### Material Design

> Most of our users are interested in the good looking functional aspect of our components, rather than in a strict application of the Material Design specification. However, we think that it's important for us to keep up to date with the guidelines. The specification received it's most important update, since Google made it the design language for its apps in 2014, at [Google I/O 2018](https://material.io/collections/google-io18/).

Reviewer: ```suggestion
Most of our users are interested in the good looking functional aspect of our components, rather than in a strict application of the Material Design specification. However, we think that it's important for us to keep up to date with the guidelines. The specification received it's most significant update since Google made it the design language for its apps in 2014, at [Google I/O 2018](https://material.io/collections/google-io18/).
```

The Material Design "v2" announcement caught us by surprise when we released Material-UI v1. We have worked on upgrading our components since then. In v4, we have updated the styles of the Tab, Snackbar, Checkboxes, Radios, Switches, List, Dialog, and other components ✨.

![material1](/static/blog/material-ui-v4-is-out/material1.png)

<p class="blog-description">Dense Outlined text field</p>

![material2](/static/blog/material-ui-v4-is-out/material2.png)

<p class="blog-description">Dense Filled text field</p>

## What's new?

There are so many new things, we can't be exhaustive. Aside from what we have already announced, you will find:

> - A new **strapline** "*React components for faster and easier web development. Build your own design system, or start with Material Design.*" that better embody our mission and emphasize the [customizability potential](https://techcrunch.com/2018/05/08/google-makes-its-material-design-system-easier-to-customize/).

Reviewer: ```suggestion
- A new **strapline** "*React components for faster and easier web development. Build your own design system, or start with Material Design.*" that better embodies our mission and emphasize the [customizability potential](https://techcrunch.com/2018/05/08/google-makes-its-material-design-system-easier-to-customize/).
```

- New [page layout examples](/getting-started/page-layout-examples/).

![layout](/static/blog/material-ui-v4-is-out/layout.png)

<p class="blog-description">One example among others</p>

- [Responsive font sizes.](http://material-ui.com/customization/typography/#responsive-font-sizes)

![font-size](/static/blog/material-ui-v4-is-out/font-size.png)

- [A Box + system utility helpers](https://medium.com/material-ui/introducing-material-ui-design-system-93e921beb8df).

![box](/static/blog/material-ui-v4-is-out/box.png)

An application example of the Box component.

- [A better spacing helper.](http://material-ui.com/customization/spacing/)

> ![box](/static/blog/material-ui-v4-is-out/box.png)

Reviewer: ```suggestion
![spacing](/static/blog/material-ui-v4-is-out/spacing.png)
```
All the images need to be given useful alt text.

> customization example of the new spacing helper.

Reviewer: ```suggestion
An example of the new spacing helper API.
```

Reviewer: We should make these consistent

- A well supported [Gatsby plugin](https://github.com/hupe1980/gatsby-plugin-material-ui) for *@material-ui/styles*.
- A [Cookbook](https://www.packtpub.com/application-development/react-material-ui-cookbook) published by Packt and written by Adam Boduch.

![cookbook](/static/blog/material-ui-v4-is-out/cookbook.png)

## What's next?

Together, we have accomplished most of the objectives we defined a year ago in the v1 release [blog post](https://medium.com/material-ui/material-ui-v1-is-out-e73ce13463eb). We're proud of everyone who contributed. We're going to try to execute these new objectives with the same regularity:
- **More components.** This is, by far, the most requested improvement dimension of the library. We have identified a few useful components we are interested in building:
  - Layout components
  - AutoComplete/ ComboBox / Dropdown List / MultiSelect
  - Slider in core and the support for range values
  - Tree view
  - Dropzone / Upload
  - Skeleton
  - Jumbotron
  - Carousel
  - Rating
  - Timeline
  - Advanced table

Let us know the components you want! 🚀

> - **Documentation**. The documentation is paramount. We are working on a page raking mechanism that will help us better prioritize our effort on the documentation.

Reviewer 2: ```suggestion
- **Documentation**. The documentation is paramount. We are working on a page ranking mechanism that will help us better prioritize our effort on the documentation.
```

- **Use `system` in the core.** We have received great feedback on the new system package. Going forward, we will try to move it to the core components. Ideally, you should be able to use dynamic color & variants from your theme as well as have access to all the props:

```jsx
import { Button } from '@material-ui/core';

<Button mt={{ xs: 2, md: 3 }}>Hello worlds</Button>
```

> - **Styled components.** We have seen many people asking for a migration to styled components. We want Material-UI v5 to be better aligned with the community's best-loved tools, but at the same time, we don't want to break your code.

Reviewer: ```suggestion
- **Styled components.** We have seen many people asking for migration to styled components. We want Material-UI v5 to be better aligned with the community's best-loved tools, but at the same time, we don't want to break your code.
```
